### PR TITLE
Dynamic Resolution Tests

### PR DESCRIPTION
**Pull request by kecho**
### Purpose of this PR
* Adding new dynamic resolution tests, for software and hardware on all platforms.
* DRS can now run on render to texture cameras.
* Adding new callback for CustomPostProcess pass. This callback is required since we change the filter mid frame. Its the only possible way I found to set the filter on a per camera basis.

image

### Comments to reviewers
This PR includes a new callback in the CustomPass. If we dont wan't to expose this, please provide feedback on other possible ways we could change the filter mid frame.
Another possibility as well is to create 8 separate tests, and keep the filter on a per frame basis (not per camera basis). This however would add 8 new tests.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPassBookmarkEventContext.cs
Comment (kecho) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPassBookmarkEventContext.cs:3]: Will add proper docs here, if you guys agree on this callback arguements / naming

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: Adding @alelievr  for review on adding callback in custom pass

**Review by sebastienlagarde — APPROVED** [review]
I am ok with the change on my side, but prefer to let @alelievr  judge about the added callback or if we want to do it another way

**Comment by sebastienlagarde** [comment]
Note: you should trigger Yamato

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: @kecho I'm not sure to see why we need to add another user callback for such case, there is already the `Execute()` function of the custom pass that should be enough.

Additionally, your callback is outside of the render graph, so I'm really not sure when it would be executed.

Also, I don't understand the purpose of the enum `CustomPassRenderEventContext.EventType`. We already have the injection point available in the custom pass API: https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@10.3/api/UnityEngine.Rendering.HighDefinition.CustomPass.html#UnityEngine_Rendering_HighDefinition_CustomPass_injectionPoint

Comment (kecho) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: Hey @alelievr , the problem with Execute is that it happens too late. It happens in the execution of the render graph itself. I need a callback that happens during render graph construction, I thought having some sort of 'event' hook would make it easy to add other injection points. 
Alternatively, I need a callback that ticks once per camera, before performing any rendering.  This tick must happen during render graph construction, prior to running the render graph.
I need to specify a new upsampling filter per camera, and this must happen before render graph execution, otherwise its too late.

Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: Okay, so indeed, in that case, we don't have yet an API that provides access during the construction of the render graph. We still have the plan to expose such API when we'll move the custom passes to render graph (with full API on the user side).

So right now, I'd say that it's better to not expose the AAPI you added because we won't be able to remove it before the replacement lands (if it's public, it needs to be documented and kept alive for migration purposes during multiple major versions :/ )

Just to make sure, right now this callback is only used during our tests? If that's the case, we can either make it internal and then replace it with a proper API when it lands. Or you can use the This callback:
https://docs.unity3d.com/ScriptReference/Rendering.RenderPipelineManager-beginCameraRendering.html

It's called at the very beginning of the frame before the culling, would this work for you?

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: >I need a callback that ticks once per camera, before performing any rendering / render graphing.
We have BeginCameraRendering (https://docs.unity3d.com/2021.1/Documentation/ScriptReference/Rendering.RenderPipeline.BeginCameraRendering.html) but not sure if it fit all your need.

Comment (kecho) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: @sebastienlagarde  this delegate will work! However I will have to make  some refactors on how the state of the DynamicResolutionManager gets propagated. Will have to run it by @FrancescoC-unity  but sounds like the way to go then.

Comment (kecho) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: Did the changes, worked like a charm. @alelievr  won't need the custom pass callback. Thanks for checking in. Will wait on a review from Francesco now :)

Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:216]: Did you push said changes? I see only formatting as recent push

Comment (kecho) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs]: woops! sorry just pushed again

Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs]: Oh the change looks trivial to me, I am good with it 🟢

**Review by FrancescoC-unity — APPROVED** [review]

**Review by alelievr — APPROVED** [review]

**Review by sebastienlagarde — APPROVED** [review]

**Comment by sebastienlagarde** [comment]
Need to update screenshots now :)

**Comment by kecho** [comment]
Final screenshots have been updated. Triggering yamato one more time.

**Comment by sebastienlagarde** [comment]
Test is green, I will move forward and merge it as it can prevent to break code. If QA have feedback, please raise them, thanks.

**Review by JMargevics — COMMENTED** [review]
Checked already merged. Looks good 👍

**Pull request merged**
